src/transport/manager/mod.rs

```diff
@@ -697,6 +721,21 @@ impl TransportManager {
         ) {
             PeerState::Dialing { ref mut record } => {
                 debug_assert_eq!(record.connection_id(), &Some(connection_id));
+                if record.connection_id() != &Some(connection_id) {
```

Comment:
Isn't this triggered if we dial two or more addresses simultaneously? Basically, the case when we know more than one address of a peer and employ dialing parallelism?

Reply:
Thinking out loud, I remember the state machine being similar to:

```
fn dial_addr (single_address) -> PeerState::Dialing { record: Some( single_address) } 

fn dial () -> this is where we employ the parallel dial. All dial records share the same connection ID
   -> PeerState::Opening { records, connection_id, ...transports } 
   
              fn dial() -> multiple parallel dials (max 8)
Disconnected  ---------------------------------> Opening 
  
              on_open_sucesss: negotiate + cancel other dials
Opening   -----> Dialing { record: first successful record }
              if other dial attempts succeed before canceling, then we return InvalidState (and just log it)
              on_open_failure: lose track of the dial
Opening ------> Opening
    
```

That points me towards adding a nice state machine and documenting these states.
It gets complicated by just looking at the code and we can easily forget it.

A few other things that I could address here:
- remove debug_assert from on_open_failure on_open_sucess; since we can race with the transports
- a dial failure can come later (same as above) and on_dial_failure we have a todo if we are in `Opening`

Reply:
You are right, looked into the sources and now understand I was missing we replace the address record with the first succeeded record.